src/blocks/agc.rs

```diff
@@ -0,0 +1,257 @@
+use num_complex::ComplexFloat;
+use futuresdr_pmt::Pmt;
+use futures::FutureExt;
+
+use crate::anyhow::Result;
+use crate::runtime::Block;
+use crate::runtime::BlockMeta;
+use crate::runtime::BlockMetaBuilder;
+use crate::runtime::Kernel;
+use crate::runtime::MessageIo;
+use crate::runtime::MessageIoBuilder;
+use crate::runtime::StreamIo;
+use crate::runtime::StreamIoBuilder;
+use crate::runtime::WorkIo;
+
+
+pub struct AGC<T> {
+    // Minimum value that has to be reached in order for AGC to start adjusting gain.
+    squelch: T,
+    // maximum gain value (0 for unlimited).
+    max_gain: T,
+    // initial gain value.
+    gain: T,
+    // reference value to adjust signal power to.
+    reference_power: T,
+    // the update rate of the loop.
```

Comment:
The meaning of "update_rate" was unclear to me by name before looking at the code. I first thought you might be doing periodic amplitude sampling. Perhaps something like "slew_rate", "gain_rate", or "adjustment_rate"? Pretty minor though, ignore at will.

Also, might as well use "doc comments" "///" here.

Reply:
GNU Radio just uses the term "rate". 

https://wiki.gnuradio.org/index.php/AGC